pkg/scaffold/scaffold.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// Boilerplate allows a Template to get the Boilerplate to put at the top of go files
+type Boilerplate interface {
+	// SetBoilerplate sets the boilerplate file content
+	SetBoilerplate(string)
+}
+
+// BoilerplatePath allows a Template to get the path to the boilerplate file
+type BoilerplatePath interface {
+	// SetBoilerplatePath sets the path to the boilerplate file
+	SetBoilerplatePath(string)
+}
+
+// Options are the options for executing scaffold templates
+type Options struct {
+	// BoilerplatePath is the path to the boilerplate file
+	BoilerplatePath string
+
+	// Path is the path to the project
+	ProjectPath string
+}
+
+// Scaffold writes Templates to scaffold new files
+type Scaffold struct {
+	// BoilerplatePath is the path to the boilerplate file
+	BoilerplatePath string
+
+	// Boilerplate is the contents of the boilerplate file for code generation
+	Boilerplate string
+
+	BoilerplateOptional bool
+
+	// Project is the project
+	Project project.Project
+
+	ProjectOptional bool
+}
+
+func (s *Scaffold) setFieldsAndValidate(t Template) error {
+	// Set boilerplate on templates
+	if b, ok := t.(BoilerplatePath); ok {
+		b.SetBoilerplatePath(s.BoilerplatePath)
+	}
+	if b, ok := t.(Boilerplate); ok {
+		b.SetBoilerplate(s.Boilerplate)
+	}
+	if b, ok := t.(Project); ok {
+		b.SetProject(s.Project)
+	}
+
+	// Validate the template is ok
+	if v, ok := t.(Validate); ok {
+		if err := v.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Scaffold) getBoilerplate(path string) (string, error) {
+	return project.GetBoilerplate(path)
+}
+
+func (s *Scaffold) getProject(path string) (project.Project, error) {
+	return project.GetProject(path)
+}
+
+func (s *Scaffold) defaultFunc(options *Options) error {
+	if options.BoilerplatePath == "" {
+		options.BoilerplatePath = project.BoilerplatePath()
+	}
+	if options.ProjectPath == "" {
+		options.ProjectPath = project.Path()
+	}
+
+	s.BoilerplatePath = options.BoilerplatePath
+
+	var err error
+	s.Boilerplate, err = s.getBoilerplate(options.BoilerplatePath)
+	if !s.BoilerplateOptional && err != nil {
+		return err
+	}
+
+	s.Project, err = s.getProject(options.ProjectPath)
+	if !s.ProjectOptional && err != nil {
+		return err
+	}
+	return nil
+}
+
+// Execute executes the Templates
+func (s *Scaffold) Execute(options Options, elem ...Template) error {
+	if err := s.defaultFunc(&options); err != nil {
+		return err
+	}
+	for _, t := range elem {
+		err := s.setFieldsAndValidate(t)
+		if err != nil {
+			return err
+		}
+
+		b, _ := ioutil.ReadFile(t.Path())
+
+		// Write the template
+		temp := newTemplate(t)
+		err = t.Execute(b, temp, func() io.WriteCloser { return newWriteCloser(t.Path()) })
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func newWriteCloser(path string) io.WriteCloser {
+	dir := filepath.Dir(path)
+	err := os.MkdirAll(dir, 0700)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fi, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0600)
```

Comment:
add `O_TRUNC` option

Reply:
Oh, I noticed func `Execute` will return directly if the file exists. So the overwrite scenario cannot happen.
But for robustness, I think we'd better have the opition here, in case it used in other places.